### ISSUE TITLE
C# Collectible .net assemblies may be breaking certain packages that rely on Castle proxies (such as Moq)

### ISSUE THREAD

**Issue by definitelyokay** [godotengine/godot#66060]
### Godot version

4.0.beta1.mono.official

### System information

macOS 12

### Issue description

TL;DR: Collectible C# assemblies are possibly breaking some very popular 3rd party nuget packages.

This one's a bit complicated, so please be patient with me.

I use [Moq](https://github.com/moq/moq4) in conjunction with my own test runner [go_dot_test](https://github.com/chickensoft-games/go_dot_test) for writing C# unit tests for 5 C# Godot packages I maintain. There are probably several hundred unit tests written using Moq across all of the packages. In Godot 3.x, everything runs fine targeting netstandard2.1.

Background: I am working on porting `go_dot_test` itself to Godot 4. It uses itself and Moq to test itself. So far, I've succeeded in getting GoDotTest's code ported over to run in Godot4, but the tests do not pass because Moq breaks with an obscure error in the .net6 environment. All of my stuff seems to work, but Moq does not.

What's breaking: Creating a new mock and grabbing the object breaks Moq:

```cs
var myMock = new Mock<IMyInterface>()
var mockObj = myMock.Object; // Throws System.NotSupportedException — A non-collectible assembly may not reference a collectible assembly.
```

I've searched around the last day or so for "A non-collectible assembly may not reference a collectible assembly" error and the only things I've found that seem relevant are these two links, which I don't fully understand.

- [Stack Overflow: A non-collectible assembly may not reference a collectible assembly](https://stackoverflow.com/questions/63077140/a-non-collectible-assembly-may-not-reference-a-collectible-assembly)
- [Microsoft: Collectible Assembly in non-collectible AssemblyLoadContext](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/7.0/collectible-assemblies)

From what I understand from the Stack Overflow post, collectible assemblies may be breaking projects which create a certain type of proxy using the Castle system. Not entirely sure as I'm way out of my depth.

If I'm reading the Godot source code correctly, Godot may actually be forcing C# assemblies to load as collectible, which may be causing the error:

https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/modules/mono/glue/GodotSharp/GodotPlugins/PluginLoadContext.cs#L17-L20

Here's a screenshot of the error, if it's helpful.

![Screen Shot 2022-09-18 at 12 17 27 PM](https://user-images.githubusercontent.com/16472250/190919986-da115d45-21c3-4ab2-bfa0-9f9591bcb9fc.png)

### Steps to reproduce

- Create a new Godot 4 project.
- Add a C# script.
- Include Moq:

```xml
<PackageReference Include="Moq" Version="4.18.2" />
```

- Use moq to create a mock object and reference `mock.Object`.

Moq will throw a `NotSupportedException` about collectible assemblies.

### Minimal reproduction project

For a minimum reproduction sample, see the demo [CollectibleAssemblies](https://github.com/chickensoft-games/CollectibleAssemblies) I created to demonstrate this issue.

You can debug the project using the launch configuration in VSCode.

You will need a GODOT4 environment variable pointing to the Godot4 executable on your machine for the VSCode `launch.json` debugger configuration to work. Or, you can edit the launch configuration in `.vscode/launch.json` to point directly to Godot 4 on your machine by replacing `${env:GODOT4}`.

**Comment by Maclay74** [comment]
I have the same problem, but using it with Harmony. It can't work with Collecible Assemblies as well.
I need it to patch _Ready method and inject dependencies there.

**Comment by Maclay74** [comment]
> Is that a possible work-around?

I tried to implement DI in my application, so I needed to inject dependencies into nodes.
We don't have access to ctor, the closest we have is `_Ready` method. So, essentially, I hook it on IL level, call my `inject()` method and give its control back. So I used something called Harmony, which requires assembly to not be collectible. 

Building engine is not necessary, you would need to build only c# bindings.

Anyway, let's hope that this flag isn't something important and we can have it switched back eventually. As for now, the only workaround I see is building bindings. 

Sorry that it won't work in your case.

**Comment by neikeq** [comment]
The assemblies must be collectible in order for the editor to be able to reload them after building. This is the only way to reload assemblies in .NET 6, so disabling it is not an option.

Do you need to run this in the editor? If not, then we could set `isCollectible: false` when not in the editor.

**Comment by neikeq** [comment]
Sorry, for the delay. Can you check if the PR solves the issue?

**Comment by akien-mga** [comment]
Fixed by #67511.
